pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -432,56 +431,39 @@ func (m *ManagerImpl) GetCapacity() (v1.ResourceList, v1.ResourceList, []string)
 	return capacity, allocatable, deletedResources
 }
 
-// checkpointData struct is used to store pod to device allocation information
-// and registered device information in a checkpoint file.
-// TODO: add version control when we need to change checkpoint format.
-type checkpointData struct {
-	PodDeviceEntries  []podDevicesCheckpointEntry
-	RegisteredDevices map[string][]string
-}
-
 // Checkpoints device to container allocation information to disk.
 func (m *ManagerImpl) writeCheckpoint() error {
 	m.mutex.Lock()
-	data := checkpointData{
-		PodDeviceEntries:  m.podDevices.toCheckpointData(),
-		RegisteredDevices: make(map[string][]string),
-	}
+	data := NewDevicePluginCheckpoint()
+	data.(*checkpointData).podDeviceEntries = m.podDevices.toCheckpointData()
 	for resource, devices := range m.healthyDevices {
-		data.RegisteredDevices[resource] = devices.UnsortedList()
+		data.(*checkpointData).RegisteredDevices[resource] = devices.UnsortedList()
 	}
 	m.mutex.Unlock()
-
-	dataJSON, err := json.Marshal(data)
-	if err != nil {
-		return err
-	}
-	err = m.store.Write(kubeletDevicePluginCheckpoint, dataJSON)
+	err := m.checkpointManager.CreateCheckpoint(kubeletDevicePluginCheckpoint, data)
 	if err != nil {
-		return fmt.Errorf("failed to write deviceplugin checkpoint file %q: %v", kubeletDevicePluginCheckpoint, err)
+		return fmt.Errorf("failed to write checkpoint file %q: %v", kubeletDevicePluginCheckpoint, err)
 	}
 	return nil
 }
 
 // Reads device to container allocation information from disk, and populates
 // m.allocatedDevices accordingly.
 func (m *ManagerImpl) readCheckpoint() error {
-	content, err := m.store.Read(kubeletDevicePluginCheckpoint)
+	checkpoint := NewDevicePluginCheckpoint()
+	err := m.checkpointManager.GetCheckpoint(kubeletDevicePluginCheckpoint, checkpoint)
 	if err != nil {
-		if err == utilstore.ErrKeyNotFound {
+		if err == errors.ErrCheckpointNotFound {
```

Comment:
As far as I can see this error is defined and is checked for but isn't actually returned by any function.
In case of nonexistent checkpoint, `GetCheckpoint()` will return `ErrKeyNotFound` from `utilstore` package.
So I guess either `GetCheckpoint()` should check for `ErrKeyNotFound` and return `ErrCheckpointNotFound` or `ErrCheckpointNotFound` should be dropped.
In case of `deviceplugin` it doesn't make a big difference as error returned by `NewDevicePluginCheckpoint` is ignored and only logged anyway, but it may cause inconsistency and errors in other packages.

Reply:
@sjenning do you think it should be addressed before merging or rather in a follow-up PR?
CC: @ConnorDoyle

Reply:
Updated :)
And rebased as well. Thanks!!